### PR TITLE
(router): add url hash support

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

Adds [URL hash](https://developer.mozilla.org/en-US/docs/Web/API/URL/hash) support to Expo Router. This enable 3rd party services that use the hash as part of work flows (e.g authentication) 

# How

The hash is converted into a search parameter called `#`. This was chosen as its not a valid search parameter and the encoded version is unlikely to be in use.

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by marklawlor** [comment]
Blocked by #27285, which will fix the failing tests

File: packages/expo-router/src/fork/__tests__/getStateFromPath.test.web.ts [hunk 1]
```diff
@@ -100,19 +100,9 @@ describe(getUrlWithReactNavigationConcessions, () => {
...
-
-  [
-    ['', ''],
-    ['https://acme.com/hello/world/?foo=bar#123', 'https://acme.com/hello/world/?foo=bar'],
-    ['/foobar#123', '/foobar'],
-  ].forEach(([url, expected]) => {
-    it(`returns the pathname without hash for ${url}`, () => {
-      expect(getUrlWithReactNavigationConcessions(url).inputPathnameWithoutHash).toBe(expected);
-    });
-  });
```
Comment (marklawlor) [inline packages/expo-router/src/fork/__tests__/getStateFromPath.test.web.ts:112]: These were removed as `inputPathnameWithoutHash` isn't actually used in the code base and we not support hashes

**Review by amandeepmittal — APPROVED** [review]
Docs changes lgtm 👍

File: docs/pages/router/reference/search-parameters.mdx [hunk 2]
```diff
@@ -251,3 +251,25 @@ export default function User() {
   );
 }
 ```
+
+## Hash support
+
+The URL (hash)[https://developer.mozilla.org/en-US/docs/Web/API/URL/hash] is a string that follows the `#` symbol in a URL. It is commonly used on websites to link to a specific section of a page, but it can also be used to store data. Expo Router treats the hash as a special search parameter using the name `#`. It can be accessed and modified using the same hooks and APIs from [search parameters](/router/reference/search-parameters/).
+
+```tsx app/hash.tsx
```
Comment (EvanBacon) [inline docs/pages/router/reference/search-parameters.mdx:259]:
```suggestion
```tsx app/index.tsx
```

**Comment by marklawlor** [comment]
Comment out docs so they are not deployed

**Review by EvanBacon — APPROVED** [review]

**Pull request merged**